walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -22,35 +22,57 @@
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
     Collection<String> attendees = request.getAttendees();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
     PriorityQueue<TimeRange> blockingEvents = new PriorityQueue<>(TimeRange.ORDER_BY_START);
+    PriorityQueue<TimeRange> blockingEventsNonOptional =
+        new PriorityQueue<>(TimeRange.ORDER_BY_START);
     for (Event event : events) {
       if (!Collections.disjoint(event.getAttendees(), attendees)) {
+        blockingEventsNonOptional.add(event.getWhen());
         blockingEvents.add(event.getWhen());
+      } else {
+        if (!Collections.disjoint(event.getAttendees(), optionalAttendees)) {
+          blockingEvents.add(event.getWhen());
+        }
       }
     }
     Collection<TimeRange> freeSlots = new ArrayList<>();
     long meetingDuration = request.getDuration();
     if (meetingDuration <= TimeRange.WHOLE_DAY.duration()) {
       if (!blockingEvents.isEmpty()) {
-        int previousEventEnd = TimeRange.START_OF_DAY;
-        while (!blockingEvents.isEmpty()) {
-          TimeRange event = blockingEvents.remove();
-          int gapDuration = event.start() - previousEventEnd;
-          if (gapDuration >= meetingDuration) {
-            freeSlots.add(TimeRange.fromStartEnd(previousEventEnd, event.start(), false));
-          }
-          if (event.end() > previousEventEnd) {
-            previousEventEnd = event.end();
+        findFreeSlots(blockingEvents, meetingDuration, freeSlots);
+        if (freeSlots.isEmpty()) {
+          if (!blockingEventsNonOptional.isEmpty()) {
```

Comment:
(nit) It's slightly more readable if you say
if (statement is true) ... else ...
rather than
if(not(statement is true)) ... else